Reimplement 'pause' in C - smaller footprint all around

uluyol opened the pull request:
Statically links against musl. Size of amd64 binary is 3560 bytes.

I couldn't test the arm binary since I have no hardware to test it on, though I assume we want it to work on a raspberry pi.

This PR also adds the gcc5/musl cross compiling image used to build the binaries.

@thockin

uluyol commented on build/musl-cross/Makefile:
```diff
@@ -14,14 +12,13 @@
-set -e
-set -x
+REPO = docker.io/uluyol/cross
+TAG = musl-1.1.14
```
REPO needs to be changed before submission along with docker push command.

uluyol commented on build/pause/Makefile:
```diff
@@ -12,20 +12,30 @@
-# IMAGE = pause-arm
+TRIPLE = x86_64-linux-musl
+IMAGE = docker.io/uluyol/pause
+TAG = 0.1
```
REPO and TAG need to be changed before submission along with docker push command.

luxas replied:
It should be `gcr.io/google-containers/pause-${ARCH}` and maybe `3.0`

uluyol commented on build/pause/Makefile:
```diff
@@ -12,20 +12,30 @@
-	./prepare.sh ${ARCH}
-	docker build -t gcr.io/google_containers/$(IMAGE):$(TAG) .
+pause: pause.c
+	docker run -v $(CURDIR):/build docker.io/uluyol/cross:musl-1.1.14 \
```
Use gcr.io image once it's available.

uluyol commented:
I wasn't sure how (if at all) to integrate building these with any other part of the build system.

ihmccreery commented:
Reassigning to @thockin for triage.  I'll ask, though: why?

uluyol commented:
Should have linked this earlier: #15140

uluyol commented:
@k8s-merge-robot you're a little too eager

luxas commented:
@uluyol Thanks, I'll check this out.

luxas commented:
First off, why do this in musl? It should be statically linked anyway.
Also it would be good to not use a new `musl-cross` image but `kube-cross`.

How much cpu/ram does this consume? 1 RSS? The thing with C should be to reduce the usage to a minimum.
I would also be great to get the binary slighty smaller. 
Quoting @thockin 
> a C program (I just did it) has an RSS of 1 page. The cost is, as you point out, you need a C compiler to rebuild it. If you want to make the binary tiny you also need dietlibc.

Maybe dietlibc would be something to look into?

uluyol commented:
> First off, why do this in musl? It should be statically linked anyway.
Also it would be good to not use a new musl-cross image but kube-cross.

glibc doesn't support static linking and isn't exactly lightweight. Does kube-cross have a C cross-compiler? I was under the impression that it's just for Go.

> How much cpu/ram does this consume? 1 RSS? The thing with C should be to reduce the usage to a minimum.
I would also be great to get the binary slighty smaller. 

The binary is already smaller than 4KB, I haven't measured the memory usage but the entire binary could fit in a single page.

> Maybe dietlibc would be something to look into?

I don't see any advantage to using dietlibc over musl. musl works with kubernetes dns for things that care now and is MIT licensed (dietlibc is GPL). Even if we use dietlibc, we still need a dedicated cross-compiler setup for it so it wouldn't simplify the build process at all.

luxas commented:
Thanks. 
Sorry I read the binary size wrongly (thought it was 3608 kB). Of course 4KB is very good.
The `kube-cross` image is [here](https://github.com/kubernetes/kubernetes/blob/master/build/build-image/cross/Dockerfile). I think it would be a better place to put it in, but it may just be me.

luxas commented on build/musl-cross/arm/config.sh:
```diff
@@ -0,0 +1,26 @@
+GCC_CONFFLAGS="--with-arch=armv5 --with-float=softfp"
+
+# Enable this to build the bootstrap gcc (thrown away) without optimization, to reduce build time
+GCC_STAGE1_NOOPT=1
```
I think we always want to optimize the binary, although it takes a little more time. 
The compilation will only be done one or twice, but the image will be used many thousands of times

uluyol replied:
Building gcc requires building it twice (first build the compiler, then use it to build itself). This disables optimizations when building the throwaway compiler.

luxas replied:
Thanks. Noticed that when I tried out the code later on.

luxas commented on build/pause/Makefile:
```diff
@@ -12,20 +12,30 @@
+			/opt/cross/$(TRIPLE)/bin/$(TRIPLE)-strip $@"
 
-push:	build
-	gcloud docker --server=gcr.io push gcr.io/google_containers/$(IMAGE):$(TAG)
```
This should be kept as-is

uluyol replied:
Done.

luxas commented:
@uluyol FYI, there should be an image for `ppc64le` and `arm64` too

luxas commented on build/pause/Makefile:
```diff
@@ -12,20 +12,30 @@
+CFLAGS = -O2 -Wall -static
 
-TAG = 2.0
+#TRIPLE = arm-linux-musleabi
```
Please do: 
```
ARCH?=amd64
ifeq($(ARCH),amd64)
TRIPLE?=x86_64-linux-musl
endif
ifeq($(ARCH),arm)
TRIPLE?=arm-linux-musleabi
endif
ifeq($(ARCH),arm64)
TRIPLE?=aarch64-linux-musl
endif
ifeq($(ARCH),ppc64le)
TRIPLE?=powerpc-linux-musl # Not sure about the naming here
endif
```
One shouldn't have to edit the file to cross-build, just specify args

uluyol replied:
Done.

uluyol commented:
musl (and dietlibc) don't have support for ppc64 or arm64 so those continue to use Go.

uluyol commented:
Container sizes:

    pause-arm           4.856 kB
    pause-amd64         3.56 kB
    pause-ppc64le       1.338 MB
    pause-arm64         1.269 MB

thockin commented on build/pause/pause.c:
```diff
@@ -0,0 +1,33 @@
+#include <unistd.h>
+
+static void sigdown(int signo) {
+    exit(0);
```
log the signal received

thockin commented on build/pause/pause.c:
```diff
@@ -0,0 +1,33 @@
+    exit(0);
+}
+
+int main(int argc, const char **argv) {
```
if we don't use args, no need to spec them

thockin commented:
I don't understand the comment about glibc not supporting static compilation, which leads to a right MESS of things.  On what planet does compiling a trivial C program with `-static` not work?

```
thockin@thockin-glaptop2:~/Downloads$ md5sum pause.c 
65ae4962d8f9baf8ff6c15944f3dacd7  pause.c

thockin@thockin-glaptop2:~/Downloads$ gcc -o pause -Os -static pause.c

thockin@thockin-glaptop2:~/Downloads$ ls -l pause
-rwxr-x--- 1 thockin eng 877289 Mar 20 22:00 pause

thockin@thockin-glaptop2:~/Downloads$ ./pause &
[2] 11877

thockin@thockin-glaptop2:~/Downloads$ ps auxw | head -1
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND

thockin@thockin-glaptop2:~/Downloads$ ps auxw | grep ./pause
thockin  11877  0.0  0.0   1072     4 pts/22   S    22:00   0:00 ./pause

thockin@thockin-glaptop2:~/Downloads$ kill -TERM %1

thockin@thockin-glaptop2:~/Downloads$ 
[1]-  Done                    ./pause

thockin@thockin-glaptop2:~/Downloads$ 
```

The binary is larger than I am happy about but it's pretty inconsequential, and still smaller than the Go binary.

thockin commented on build/pause/pause.c:
```diff
@@ -0,0 +1,33 @@
+    if (signal(SIGINT, sigdown) == SIG_ERR)
+        return 1;
+    if (signal(SIGTERM, sigdown) == SIG_ERR)
+        return 1;
```
return 2 here

thockin commented on build/pause/pause.c:
```diff
@@ -0,0 +1,33 @@
+        return 1;
+    signal(SIGKILL, sigdown);
+    for (;;) pause();
+    return 1;
```
return 42 (or something big but <128) here.

thockin replied:
maybe printf("impossible - the infinite loop terminated!")

luxas commented on build/pause/Makefile:
```diff
@@ -12,20 +12,54 @@
+	USEGO = 1
+endif
+
+IMAGE = docker.io/uluyol/pause-$(ARCH)
```
Please update this now so we see how this is going to look like when ready

luxas commented:
@uluyol I think we should keep `musl-cross` (if we want it), not integrate it with `kube-cross` as I thought earlier.
I may then do a follow-up PR that uses `kube-cross` for the `go` compilation and also `upx`es the binary inside docker.

luxas commented on build/musl-cross/Dockerfile:
```diff
@@ -0,0 +1,35 @@
+	git clone https://github.com/sabotage-linux/musl-cross.git && \
+	cd musl-cross && git checkout a4777a6
+
+ADD amd64 /cross-build/conf/amd64
```
I think it's clearer if you rename `amd64/config.sh` to `config_amd64.sh` and `ADD config_amd64.sh /cross-build/conf/amd64/config.sh`. No extra dirs are created then

uluyol commented:
>I don't understand the comment about glibc not supporting static compilation, which leads to a right MESS of things. On what planet does compiling a trivial C program with -static not work?

Sorry, I've seen a number of posts discouraging this, turns out it's only an [issue if you use NSS](https://sourceware.org/glibc/wiki/FAQ#Even_statically_linked_programs_need_some_shared_libraries_which_is_not_acceptable_for_me.__What_can_I_do.3F). I'll use glibc for the other architectures.

thockin commented:
Just do the same thing for them all.  The size is sooooo not a big deal.

On Mon, Mar 21, 2016 at 9:49 PM, Muhammed Uluyol <notifications@github.com>
wrote:

> I don't understand the comment about glibc not supporting static
> compilation, which leads to a right MESS of things. On what planet does
> compiling a trivial C program with -static not work?
>
> Sorry, I've seen a number of posts discouraging this, turns out it's only
> an issue if you use NSS
> <https://sourceware.org/glibc/wiki/FAQ#Even_statically_linked_programs_need_some_shared_libraries_which_is_not_acceptable_for_me.__What_can_I_do.3F>.
> I'll use glibc for the other architectures.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23009#issuecomment-199635845>
>

uluyol commented:
Got rid of musl-cross since we use glibc everywhere now. I don't see any references to ppc64le or arm64 in kube-cross, so this only builds for amd64 and arm.

luxas commented on build/pause/Makefile:
```diff
@@ -12,20 +12,35 @@
 
-all:	push
+pause: pause.c
+	docker run -v $(CURDIR):/build gcr.io/google_containers/kube-cross:v1.4.2-1 \
```
Break out `v1.4.2-1` to a variable above

thockin replied:
+1

uluyol replied:
See update, it reads this from build-all/cross/VERSION

luxas commented:
This starts to look really good. 
I'll implement `arm64` and `ppc64le` as soon as we have support for it (when golang 1.6)

luxas commented:
@thockin While your at it, please review #21617 too ASAP. 
Will only do a small rebase before it's ready to merge

thockin commented on build/pause/Makefile:
```diff
@@ -12,20 +12,35 @@
+			$(TRIPLE)-gcc $(CFLAGS) -o pause $< && \
+			$(TRIPLE)-strip pause"
+
+build: pause
```
s/build/container - I think that is what other sub-modules are doing

uluyol replied:
Done.

thockin commented on build/pause/Makefile:
```diff
@@ -12,20 +12,35 @@
+build: pause
+	docker build -t $(IMAGE):$(TAG) .
+
+push: build
```
we should probably start setting a better example.  This is a bit clunky.  We should do something like:

```
container .container: $(BIN)
  X=$$(docker build ...) &&  echo $$X > .container

push: .container
  docker push ...
```

don't forget to clean up .container

thockin replied:
might also add a .push file so you don't push if you don't need to

uluyol replied:
Done.

thockin commented on build/pause/Makefile:
```diff
@@ -12,20 +12,35 @@
+TAG = 3.0
 
-all:	push
+pause: pause.c
```
BIN := pause
SRCS := pause.c

$(BIN): $(SRCS)

uluyol replied:
Done.

thockin commented on build/pause/pause.c:
```diff
@@ -16,18 +14,23 @@ See the License for the specific language governing permissions and
-	c := make(chan os.Signal, 1)
-	signal.Notify(c, os.Interrupt, os.Kill, syscall.SIGTERM)
+static void sigdown(int signo) {
+	psignal(signo, "shutting down, got signal");
```
consistent tabs and spaces.  Use tabs for indent

uluyol replied:
Done.

thockin commented:
I need to think about any ramifications to this, but I'd like to see another commit in this PR that changes the codebase to use the new tag and image name.  Do you think you can do that part, too?

luxas commented:
If not @uluyol wants to do it, I may rename references to pause.

Another important thing, we really should decide about the naming. When we're aiming for cross-platform support, image names should be consistent. This far, all cross images has the `binary-arch:version` naming which I prefer strongly. I think now when we're making a new image and updating all refs, the amd64 image name should be: `gcr.io/google_containers/pause-amd64:3.0`. Certainly `gcr.io/google_containers/pause:3.0` should also be pushed for compability. This naming would make everything easier, just write `gcr.io/google_containers/pause-${ARCH}:${VERSION}` in a script and you're done. It would workaround hacks like: #23059 (please also read the conversation there).

CCing people involved in the discussion: @mikedanese @vishh @david-mcmahon

uluyol commented:
PTAL, I added a autogenerated commit which updates existing references to pause images (there are references to several different versions) and a followup after running hack/update-all.sh. I did not update references in examples or docs even though there were a few of these.

Script used to rewrite refs:
```
#!/usr/bin/env bash
find build cmd hack pkg test -type f | grep -v '^.git' | while read p; do
	sed -i \
		-e 's|gcr.io/google_containers/pause:2.0|gcr.io/google_containers/pause:3.0|g' \
		-e 's|kubernetes/pause:go|gcr.io/google_containers/pause:3.0|g' \
		-e 's|kubernetes/pause:1|gcr.io/google_containers/pause:3.0|g' \
		-e 's|kubernetes/pause|gcr.io/google_containers/pause:3.0|g' \
		"$p"
done
```

uluyol commented:
Forgot the -amd64 suffix on the new images. Will fix.

uluyol commented:
PTAL

mikedanese commented:
I would guess conservatively that linux amd64 is >98% of our target deployments. I'm opposed to littering -amd64 everywhere. I also think that if we do tag our images amd64, it should be in the tag not in the image name.

luxas commented on build/pause/Makefile:
```diff
@@ -12,20 +12,41 @@
+			$(TRIPLE)-strip $@"
+
+container .container: pause
+	docker build -t $(IMAGE):$(TAG) .
```
maybe it's better to use a `REGISTRY?=gcr.io/google_containers` variable instead
Also, the `amd64` image maybe should be tagged `gcr.io/google_containers/pause:3.0` for backwards compability. @thockin WDYT?

See example:
https://github.com/kubernetes/kubernetes/blob/master/cluster/images/hyperkube/Makefile

luxas commented:
@mikedanese I really don't see it as littering. I agree that the vast majority of deployments will be `amd64`, of course. 

**But that doesn't stop us being clear with which arch an image was built for** (as @vishh also said). We should keep the backwards compability for a while, though. 

Kubernetes for `linux/ppc64le` might have commercial interest too (available when go1.6), so we really shouldn't close any doors for new possibilities, especially not when talking about a such simple thing as naming

zmerlynn commented:
I have a funny view: Why are we talking about keeping the cross-platform tags in the same registry path? In something like `gcr.io`, we could just have `google-containers-ppc` if we really wanted (and potentially even handle the ACLs on it differently). But I know the ship is moving the other way, and that may be heavier weight in certain cases, but for us it's actually much easier.

Pensu commented:
@zmerlynn That actually sounds like a good idea, having a separate registry for the arch. Not sure, what's it gonna broke, but I would be interested in it!

luxas commented:
@zmerlynn @Pensu But I think we will settle on `binary-arch:version`, because we would have to move so many images to `gcr.io/google_containers_amd64` if we go for `registry-arch/binary:version`

If we want to incorporate @zmerlynn's idea anyway, it might be possible to store [the core Kubernetes images listed here](https://github.com/kubernetes/kubernetes/pull/23059#issuecomment-202024763) in something like `gcr.io/kubernetes_images/`

luxas commented:
@uluyol The line in `pkg/kubelet/types/constants.go` should be removed. 
And `cmd/kubelet/app/options/options.go` should be changed like this:
```go
const (
	defaultRootDir             = "/var/lib/kubelet"
	experimentalFlannelOverlay = false

	defaultPodInfraContainerImageName    = "gcr.io/google_containers/pause"
	defaultPodInfraContainerImageVersion = "2.0"
)

// Returns the arch-specific pause image that kubelet should use as the default
func GetDefaultPodInfraContainerImage() string {
	if runtime.GOARCH == "amd64" {
		return defaultPodInfraContainerImageName + ":" + defaultPodInfraContainerImageVersion
	} else {
		return defaultPodInfraContainerImageName + "-" + runtime.GOARCH + ":" + defaultPodInfraContainerImageVersion
	}
}
```
```go
const (
	defaultRootDir             = "/var/lib/kubelet"
	experimentalFlannelOverlay = false

	defaultPodInfraContainerImageName    = "gcr.io/google_containers/pause"
	defaultPodInfraContainerImageVersion = "3.0"
)

// Returns the arch-specific pause image that kubelet should use as the default
func GetDefaultPodInfraContainerImage() string {
	return defaultPodInfraContainerImageName + "-" + runtime.GOARCH + ":" + defaultPodInfraContainerImageVersion
}
```

uluyol commented:
Has the naming convention been decided? It seems like it's image:version for amd64 and image-ARCH:version for everything else but I want to make sure before rebasing.

luxas commented:
Hi @uluyol!
The naming convention is `binary-arch:version` always.
So it should be `pause-amd64:3.0`

See my previous comment how to make update the default pause image: https://github.com/kubernetes/kubernetes/pull/23009#issuecomment-203345320

luxas commented:
Now you should build the pause image for `arm64` and `ppc64le` too
See #23931 for more info

uluyol committed:
build/pause: write in C

Builds statically against glibc. References to the old pause
image have been updated.

uluyol commented:
I'm glad amd64 isn't treated as an exception. Pushed a new version. First commit contains the Go -> C conversion, second is mostly autogenerated with the change luxas requested, third is autogenerated.

These are the remaining references to pause images which have not been changed:
```
./docs/devel/developer-guides/vagrant.md:285:    fa0e29c94501        kubernetes/pause:latest   "/pause"               8 minutes ago       Up 8 minutes        0.0.0.0:8080->80/tcp     k8s--net.a90e7ce4--7813c8bd_-_3ffe_-_11e4_-_9036_-_0800279696e1.etcd--7813c8bd_-_3ffe_-_11e4_-_9036_-_0800279696e1--baf5b21b
./examples/fibre_channel/README.md:71:090ac457ddc2        kubernetes/pause                       "/pause"            12 minutes ago      Up 12 minutes                           k8s_fcpd-rw.aae720ec_fcpd_default_4024318f-4121-11e5-a294-e839352ddd54_99eb5415   
./examples/fibre_channel/README.md:72:5e2629cf3e7b        kubernetes/pause                       "/pause"            12 minutes ago      Up 12 minutes                           k8s_fcpd-ro.857720dc_fcpd_default_4024318f-4121-11e5-a294-e839352ddd54_c0175742   
./examples/fibre_channel/README.md:73:2948683253f7        gcr.io/google_containers/pause:0.8.0   "/pause"            12 minutes ago      Up 12 minutes                           k8s_POD.7be6d81d_fcpd_default_4024318f-4121-11e5-a294-e839352ddd54_8d9dd7bf       
./examples/glusterfs/glusterfs-pod.json:12:                "image": "kubernetes/pause",
./examples/iscsi/iscsi.json:11:                "image": "kubernetes/pause",
./examples/iscsi/iscsi.json:21:                "image": "kubernetes/pause",
./examples/iscsi/README.md:97:f855336407f4        kubernetes/pause                       "/pause"            6 minutes ago       Up 6 minutes                            k8s_iscsipd-ro.d130ec3e_iscsipd_default_f527ca5b-6d87-11e5-aa7e-080027ff6387_5409a4cb
./examples/iscsi/README.md:98:3b8a772515d2        kubernetes/pause                       "/pause"            6 minutes ago       Up 6 minutes                            k8s_iscsipd-rw.ed58ec4e_iscsipd_default_f527ca5b-6d87-11e5-aa7e-080027ff6387_d25592c5
./examples/rbd/rbd-with-secret.json:12:                "image": "kubernetes/pause",
./examples/rbd/rbd.json:12:                "image": "kubernetes/pause",
./docs/user-guide/node-selection/pod-with-node-affinity.yaml:28:    image: gcr.io/google_containers/pause:2.0
./hack/testdata/multi-resource-json-modify.json:46:           "image": "gcr.io/google-containers/pause:2.0",
./hack/testdata/multi-resource-json.json:44:           "image": "gcr.io/google-containers/pause:2.0",
./hack/testdata/multi-resource-list-modify.json:50:                    "image": "gcr.io/google-containers/pause:2.0",
./hack/testdata/multi-resource-list.json:48:                    "image": "gcr.io/google-containers/pause:2.0",
./hack/testdata/multi-resource-rclist-modify.json:29:                    "image": "gcr.io/google-containers/pause:2.0",
./hack/testdata/multi-resource-rclist-modify.json:63:                "image": "gcr.io/google-containers/pause:2.0",
./hack/testdata/multi-resource-rclist.json:29:                    "image": "gcr.io/google-containers/pause:2.0",
./hack/testdata/multi-resource-rclist.json:63:                "image": "gcr.io/google-containers/pause:2.0",
./hack/testdata/multi-resource-yaml-modify.yaml:30:        image: gcr.io/google-containers/pause:2.0
./hack/testdata/multi-resource-yaml.yaml:28:        image: gcr.io/google-containers/pause:2.0
./hack/testdata/pod-apply.yaml:10:    image: gcr.io/google-containers/pause:2.0
./hack/testdata/pod.yaml:10:    image: gcr.io/google-containers/pause:2.0
```

I was especially unsure of changing the testdata references since they refer to `google-containers` not `google_containers`.

thockin commented on build/pause/Makefile:
```diff
@@ -12,42 +12,65 @@
-TAG=2.0
-REGISTRY?="gcr.io/google_containers"
+REGISTRY ?= "gcr.io/google_containers"
+IMAGE = gcr.io/google_containers/pause-$(ARCH)
```
should this be `$(REGISTRY)/pause-$(ARCH)`

uluyol replied:
Done.

thockin commented on build/pause/Makefile:
```diff
@@ -12,42 +12,65 @@
 
-TAG=2.0
-REGISTRY?="gcr.io/google_containers"
+REGISTRY ?= "gcr.io/google_containers"
```
quotes not needed

uluyol replied:
Done.

thockin commented on build/pause/Makefile:
```diff
@@ -12,42 +12,65 @@
 
+TEMP_DIR := $(shell mktemp -d)
+
+all: $(TEMP_DIR)/.push $(TEMP_DIR)/.push-legacy-too
```
Why is this all in TEMP_DIR?  This means that every `make` will be foreced to repeat the work because TEMP_DIR will be different.  Why not just put .push in the current dir and .gitignore it ?

uluyol replied:
This was to avoid pushing a (e.g.) arm image with a amd64 binary. Changed to use .build-ARCH subdirectories instead.

thockin commented on build/pause/Makefile:
```diff
@@ -12,42 +12,65 @@
+
+$(TEMP_DIR)/$(BIN): $(SRCS)
+	cp $(SRCS) Dockerfile $(TEMP_DIR)
+	docker run -v $(TEMP_DIR):/build gcr.io/google_containers/kube-cross:$(KUBE_CROSS_VERSION) \
```
variable for `gcr.io/google_containers/kube-cross` please

uluyol replied:
Done.

thockin commented on build/pause/Makefile:
```diff
@@ -12,42 +12,65 @@
 
-push: build
+push $(TEMP_DIR)/.push: $(TEMP_DIR)/.container
 	gcloud docker --server=gcr.io push $(REGISTRY)/pause-$(ARCH):$(TAG)
```
vartiable for gcr.io ?

luxas replied:
I don't think it's required to specify `--server=gcr.io` at all, since it isn't there in the most other Makefiles

uluyol replied:
Removed.

thockin commented:
My nits are all around Makefile.  Needs rebase and I will be happy to see this go in.  Thanks!

luxas commented:
I guess we want to validate it in 3-4 v1.3 alpha releases, so we should get this in now.
@uluyol Please update :smile:

uluyol commented:
I think I addressed the comments. PTAL.

thockin commented on CHANGELOG.md:
```diff
@@ -2,6 +2,7 @@
 
 - [v1.3.0-alpha.3](#v130-alpha3)
   - [Downloads](#downloads)
+- [v1.3.0-alpha.3](#v130-alpha3)
```
why are you editing CHANGELOG ?

thockin commented on build/pause/Makefile:
```diff
@@ -12,42 +12,71 @@
-ARCH?=amd64
-GOLANG_VERSION=1.6
-TEMP_DIR:=$(shell mktemp -d)
+ARCH ?= amd64
```
can you add a var that is a list of supported ARCH values since you have to un-edit CHANGELOG :)

uluyol replied:
Not sure I understand the purpose of this.

thockin replied:
I was thinking a `make containers` rule would loop for all supported architectures, and a `make pushes` would push them all.  Bad idea?

uluyol replied:
Done, though I named the rules all-container and all-push.

thockin commented:
Looks great!  Just a nit and I think you don't mean to edit CHANGELOG

thockin commented on build/pause/.gitignore:
```diff
@@ -0,0 +1 @@
+.build-*
```
This should be /.build-* - the leading slash means "only in this dir"

thockin commented on build/pause/Makefile:
```diff
@@ -12,42 +12,71 @@
 
+BUILD_DIR := $(CURDIR)/.build-$(ARCH)
+
+all: $(BUILD_DIR)/.push $(BUILD_DIR)/.push-legacy-too
```
The default rule should not push.  This is why we usually call the rules "build" and "push".

thockin replied:
Add a comment here, about this, please

thockin commented:
Actually, I have a Makefile patch for you - something smelled wrong.  soon

thockin commented:
It got sort of big, sorry.  Pasting it here.  It should apply cleanly to this tree as of right now.  Once it's merged into your PR, I can actually build and push the images and we can get this PR in.

[pr-23009.diff.txt](https://github.com/kubernetes/kubernetes/files/248049/pr-23009.diff.txt)

uluyol commented:
Applied the changes, this is definitely cleaner. PTAL

thockin commented on build/pause/.gitignore:
```diff
@@ -0,0 +1,3 @@
+/.container-*
+/.push.*
```
should be push-* ?

uluyol replied:
Done.

thockin commented:
nit on gitignore.

please squash

thockin commented:
I'm going to build and push these to see if e2e needs that to run properly

thockin commented:
One more bugfix:

```
diff --git a/build/pause/Makefile b/build/pause/Makefile
index 431e3e6..78f5917 100644
--- a/build/pause/Makefile
+++ b/build/pause/Makefile
@@ -49,7 +49,7 @@ endif
 # If you want to build AND push, see the 'push' rule.
 all: container
 
-$(BIN)-$(ARCH): bin/$(BIN)-$(ARCH)
+build: bin/$(BIN)-$(ARCH)
 
 bin/$(BIN)-$(ARCH): $(SRCS)
        mkdir -p bin
@@ -60,7 +60,7 @@ bin/$(BIN)-$(ARCH): $(SRCS)
                        $(TRIPLE)-strip $@"
 
 container: .container-$(ARCH)
-.container-$(ARCH): $(BIN)-$(ARCH)
+.container-$(ARCH): bin/$(BIN)-$(ARCH)
        docker build -t $(IMAGE):$(TAG) --build-arg ARCH=$(ARCH) .
 ifeq ($(ARCH),amd64)
        docker tag -f $(IMAGE):$(TAG) $(LEGACY_AMD64_IMAGE):$(TAG)
```

This stops 'push' from doing extra work

thockin commented:
I have built and pushed these.  When you repush with last fixes, we'll see about e2e

luxas commented on build/pause/Makefile:
```diff
@@ -12,42 +12,85 @@
+	touch $@
 
-push-legacy-too: push
+push-legacy: .push-legacy-$(ARCH)
```
@thockin I'm not sure if we should push legacy of this image anymore, as we've updated all refs to `-amd64:3.0` and the `2:0` still exists.
My point is, since the "old" naming is deprecated, it might be best to not push/refer to the old naming at all, to reduce confusion. However, if you've already pushed it, it's okay
WDYT?

thockin replied:
I'd prefer to just leave it for a while.

luxas replied:
okay, fine

uluyol commented:
@thockin e2e has passed. PTAL

thockin commented:
LGTM thanks!

uluyol commented:
k8s-merge-robot has asked k8s-bot to test this three times now. Something wrong with the merge process?
